rfc/user-secrets.md

```diff
@@ -0,0 +1,186 @@
+Each URI is structured as `secret://secret-engine-id?param1=value1&param2=value2&param3=value3` where _secret-engine-id_ is the `SecretEngine` identifier and the parameter key/value pairs are specific to each engine.
+`SecretEngine` is updated with additional `decrypt()` and `validate()` methods for `UserSecretReference` values.
+Each engine is configured to make requests for secrets using a single Spinnaker identity, typically the workload identity of Spinnaker itself for ease of integration with cloud-specific APIs.
+To demonstrate, suppose we have a secret named `my-spinnaker-secret` in AWS Secrets Manager in the `us-west-2` region under the AWS account id `222211110000` while Spinnaker is running in EKS in the `us-east-1` region under the account id `111122223333`.
+This secret would have an [ARN](https://docs.aws.amazon.com/general/latest/gr/aws-arns-and-namespaces.html) of `arn:aws:secretsmanager:us-west-2:222211110000:secret:my-spinnaker-secret`.
+Then this secret can be made available to Spinnaker by attaching an IAM policy to this secret such as:
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [{
+    "Effect": "Allow",
+    "Action": ["secretsmanager:GetSecretValue"],
+    "Resource": ["arn:aws:secretsmanager:us-west-2:222211110000:secret:my-spinnaker-secret"]
+  }]
+}
+```
+
+Then the URI `secret://secrets-manager?r=us-east-1&s=arn:aws:secretsmanager:us-west-2:222211110000:secret:my-spinnaker-secret&k=secret-key` will form the basis for user secret references to this secret.
+Granting access in this direction allows Spinnaker to read the secret, but this is insufficient for providing access controls within Spinnaker around who is allowed to use this secret.
+These access controls must instead be provided via the secret engine, typically by embedding the Spinnaker roles that are allowed to use the secret.
+The `UserSecretMapper` class provides serialization and deserialization functionality for different types of user secrets.
+At minimum, a `UserSecret` is a tree-like structure where the root object has a `type` field for indicating the type of secret.
+This type corresponds to different structures for user secrets along with some way to check that a collection of roles are all allowed to use its data.
+
+The initial type introduced is the `opaque` type which has the following fields to define.
+
+* `roles`: a list of Fiat roles that are allowed to use the data in this secret.
+Permission checks involving this list of roles must check that the set of target resource roles are all contained in this list.
+* `data`: a map of keys to base64-encoded (basic variant) binary data.
+* `stringData`: a map of keys to utf8-encoded string data.
+
+For example, the following is a snippet of a secret named `eks-sa` in `us-east-1` in the AWS account id `123455432100` encoded as JSON:
+
+```json
+{
+  "type": "opaque",
+  "roles": ["admin", "sre"],
```

Comment:
From a privilege escalation:  Is this an "OR" Or "AND" operation.  E.g. a user must be in ALL these groups to read this secret, or ONE of these groups.  That comes into delegation concerns mostly on who can update a secret, and the potential transitive impacts, particularly on account operations.  NORMALLY for example pipelines to update it you must be a member of ALL the roles in a pipeline.  i could see needing a similar constraint on secret access.  But... maybe not as secrets are a bit different in downstream impacts.

Reply:
Only have to be in one of the groups because we do time-of-use checks as well.